File: mobile/android/app/src/main/java/com/siming/mobile/data/ProjectPackageChapterState.kt
package com.siming.mobile.data

import java.time.Instant
import java.time.LocalDateTime
import java.time.OffsetDateTime
import java.time.ZoneOffset
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull

/** The same v1 import evidence contract as ProjectPackageChapterCatalogingState on PC. */
internal class ProjectPackageChapterCatalogingState(
    snapshots: List<JsonObject>,
    summaries: List<JsonObject>,
) {
    private val snapshotsByChapter = snapshots.groupBy { it.text("chapter_id") }
    private val summariesByChapter = summaries.groupBy { it.text("chapter_id") }

    fun required(chapter: JsonObject): Boolean {
        val content = chapter.text("content")
        if (content.isBlank()) return false
        val chapterId = chapter.text("id")
        val version = (chapter["current_version"] as? JsonPrimitive)?.takeUnless { it.isString }?.intOrNull
        if (chapterId.isBlank() || version == null || version < 1) return true
        val summaryTimes = summariesByChapter[chapterId].orEmpty()
            .filter { it.text("summary_text").isNotBlank() }
            .mapNotNull { timestamp(it.text("updated_at").ifEmpty { it.text("created_at") }) }
        return snapshotsByChapter[chapterId].orEmpty().none { snapshot ->
            val snapshotVersion = (snapshot["version_number"] as? JsonPrimitive)?.takeUnless { it.isString }?.intOrNull
            val savedAt = timestamp(snapshot.text("created_at"))
            snapshotVersion == version && snapshot.text("content") == content && savedAt != null &&
                summaryTimes.any { !it.isBefore(savedAt) }
        }
    }

    private fun timestamp(value: String): Instant? =
        runCatching { OffsetDateTime.parse(value).toInstant() }.getOrNull()
            ?: runCatching { LocalDateTime.parse(value).toInstant(ZoneOffset.UTC) }.getOrNull()

    private fun JsonObject.text(field: String): String =
        (get(field) as? JsonPrimitive)?.contentOrNull.orEmpty()
}

File: backend/app/services/project_package_chapter_state.py
"""Rebuild the chapter write gate from durable v1 author-data evidence."""

from __future__ import annotations

from collections import defaultdict
from datetime import UTC, datetime
from typing import Any


def _timestamp(value: Any) -> datetime | None:
    if not isinstance(value, str) or not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    return parsed.replace(tzinfo=UTC) if parsed.tzinfo is None else parsed


class ProjectPackageChapterCatalogingState:
    """v1 omits job state; an old summary alone cannot release a newer chapter."""

    def __init__(self, snapshots: list[dict[str, Any]], summaries: list[dict[str, Any]]):
        self.snapshots: dict[str, list[dict[str, Any]]] = defaultdict(list)
        self.summaries: dict[str, list[dict[str, Any]]] = defaultdict(list)
        for row in snapshots:
            self.snapshots[str(row.get("chapter_id") or "")].append(row)
        for row in summaries:
            self.summaries[str(row.get("chapter_id") or "")].append(row)

    def required(self, chapter: dict[str, Any]) -> bool:
        content = str(chapter.get("content") or "")
        if not content.strip():
            return False
        chapter_id = str(chapter.get("id") or "")
        version = chapter.get("current_version")
        if not chapter_id or type(version) is not int or version < 1:
            return True
        summary_times = [
            timestamp
            for row in self.summaries.get(chapter_id, [])
            if str(row.get("summary_text") or "").strip()
            if (timestamp := _timestamp(row.get("updated_at") or row.get("created_at"))) is not None
        ]
        for snapshot in self.snapshots.get(chapter_id, []):
            if snapshot.get("version_number") != version or snapshot.get("content") != content:
                continue
            saved_at = _timestamp(snapshot.get("created_at"))
            if saved_at is not None and any(timestamp >= saved_at for timestamp in summary_times):
                return False
        return True
